relayer/strategies.go

```diff
@@ -1,58 +1,241 @@
+				for _, channel := range srcOpenChannels {
+					if !channel.active {
+						channel.active = true
+						go RelayUnrelayedPacketsAndAcks(src, dst, maxTxSize, maxMsgLength, channel, channels, doneChan)
 					}
 				}
 
-				time.Sleep(100 * time.Millisecond)
+				for channel := range channels {
+					channel.active = false
+					break
+				}
+
+				// Make sure we are removing channels no longer in OPEN state from the slice of open channels
+				for i, channel := range srcOpenChannels {
+					if channel.channel.State != types.OPEN {
+						srcOpenChannels[i] = srcOpenChannels[len(srcOpenChannels)-1]
+						srcOpenChannels = srcOpenChannels[:len(srcOpenChannels)-1]
+					}
+				}
 			}
 		}
 	}()
 	return func() { doneChan <- struct{}{} }, nil
 }
+
+// QueryChannelsOnConnection queries all the channels associated with a connection on the src chain.
+func QueryChannelsOnConnection(src *Chain) ([]*types.IdentifiedChannel, error) {
+	// Query the latest heights on src & dst
+	srch, err := src.ChainProvider.QueryLatestHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	// Query the list of channels for the connection on src
+	var srcChannels []*types.IdentifiedChannel
+
+	if err = retry.Do(func() error {
+		srcChannels, err = src.ChainProvider.QueryConnectionChannels(srch, src.ConnectionID())
+		return err
+	}, RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+		src.LogRetryQueryConnectionChannels(n, err, src.ConnectionID())
+	})); err != nil {
+		return nil, err
+	}
+
+	return srcChannels, nil
+}
+
+// FilterOpenChannels takes a slice of channels and adds all the channels with OPEN state to a new slice of channels.
+// NOTE: channels will not be added to the slice of open channels more than once.
+func FilterOpenChannels(channels []*types.IdentifiedChannel, openChannels []*ActiveChannel) []*ActiveChannel {
+	inSlice := false
+
+	// Filter for open channels
+	for _, channel := range channels {
+		if channel.State == types.OPEN {
+
+			// Check if we have already added this channel to the slice of open channels
+			for _, openChannel := range openChannels {
+				if channel.ChannelId == openChannel.channel.ChannelId {
+					inSlice = true
+					break
+				}
+			}
+
+			// We don't want to add channels to the slice of open channels that have already been added
+			if !inSlice {
+				openChannels = append(openChannels, &ActiveChannel{
+					channel: channel,
+					active:  false,
+				})
+			}
+
+			inSlice = false
+		}
+	}
+
+	return openChannels
+}
+
+// RelayUnrelayedPacketsAndAcks will relay all the pending packets and acknowledgements on both the src and dst chains.
+func RelayUnrelayedPacketsAndAcks(src, dst *Chain, maxTxSize, maxMsgLength uint64, srcChannel *ActiveChannel, channels chan<- *ActiveChannel, doneChan chan struct{}) {
+	// make goroutine signal its death, whether it's a panic or a return
+	defer func() {
+		channels <- srcChannel
+	}()
+
+	for {
+		select {
+		case <-doneChan:
+			doneChan <- struct{}{}
```

Comment:
Re-placing the done value on the channel is unusual. If this is necessary, add an explanatory comment.

Reply:
I removed this `doneChan` in favor of using contexts